File: 59-spiral-matrix-ii/59-spiral-matrix-ii.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] ans = new int[n][n];
        int row_start = 0 , col_start =0;
        int row_end = n , col_end = n;
        int i =0 , j =0;
        int num =1;
        while(i<row_end && j<col_end){
//             to move left to right...
            while(i<row_end && j<col_end){
                ans[i][j++]= num++;
                
            }
           row_start++; j--; num--;
            
//              to move from top to bottom..
            
            while(i<row_end && j<col_end){
                ans[i++][j]=num++;
            }
            
            col_end--; i--; num--;
            
//              to move from right to left..
            
            while(i>=row_start && j>=col_start){
                ans[i][j--] = num++;
            }
            row_end--; j++ ; num--;
            
//             to move fromt bottom to top
            while(i>=row_start && j>= col_start){
                ans[i--][j]=num++;
            }
            
            col_start++; i++; num--;
            
            
        }
        
        
        
        return ans;
    }
}

File: 59-spiral-matrix-ii/59-spiral-matrix-ii.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int num = 1;
        int i =0 , j=0;
        int row_start = 0 , col_start = 0;
        int row_end = n , col_end=n;
        
       int arr[n][n];
        while(i<row_end && j<col_end){
//              for left to right...
            while(i<row_end && j<col_end){
               arr[i][j++]=num++;
            }
            row_start++; j--; num--;
            //cout<<i<<" "<<j<<endl;
            
//              for top to bottom..
            while(i<row_end && j<col_end){
                arr[i++][j]=num++;
            }
            col_end--; i--; num--;
            
//              for right to left..
            
            while(i>=row_start && j>=col_start){
                arr[i][j--]=num++;
            }
            
            row_end--; j++; num--;
            
//              for bottom to top
            
            while(i>=row_start && j>=col_start){
                arr[i--][j]=num++;
            }
            col_start++; i++; num--;
            
            //cout<<i<<" "<<j<<endl;
            
        
        }
        
        
        
        vector<vector<int>> v;
        for(int i=0;i<n;i++){
            vector<int> temp;
            for(int j=0;j<n;j++){
               temp.push_back(arr[i][j]);
            }
            v.push_back(temp);
            
        }
        return v;
    }
};